design/design.md

```diff
@@ -0,0 +1,112 @@
+Applications?
+mithril, peras, leios
+
+What cryptographic primitives will we implement?
+I propose that we implement the generic ALBA primitive to make it as applicable
+as possible.
+This means there is only a generic notion of "elements" and no notions of
+signatures, vrf, etc.
+This would be somewhat difficult to work with for all our three applications
+that really just need a weighted multisignature scheme.
+Hence, my second proposal is to use this generic ALBA implementation to implement
+a convenient multisignature scheme.
+It would also serve as a nice example of how the generic ALBA code can be used.
+
+Which ALBA algorithms to implement?
+For the centralized unweighted setting, there are four known algorithms to date:
+(basic) Telescope, Telescope with prehashing, Telescope with prehashing,
+repetitions and bounded DFS, and finally simple lottery.
+I propose that we only implement the third scheme -- Telescope with prehashing,
+repetitions and bounded DFS, and possibly the simple lottery.
+The basic Telescope is way slower and offers almost no advantage over the other
+schemes.
+Telescope with prehashing is only applicable when the number of honest prover's
+element n_p is large which is probably uncommon.
+Moreover, it offers no advantage over the third scheme except for a very tiny
+difference in the certificate size (integer v).
+As an optimization of the Telescope with prehashing, repetitions and bounded
+DFS, we could even eliminate this integer from the certificate if the
+number of allowed repetitions r = 1.
+Finally, the simple lottery, while having large proof sizes, offers very
+fast proving times which might be important for some applications.
+For the decentralized setting and weighted settings, we again have four
+possible algorithms:
+the three Telescope algorithms combined with the simple lottery, and the
+simple lottery itself.
+I propose we implement Telescope with prehashing, repetitions and bounded DFS,
+as well as, the single simple lottery.
+Similar to above, the other Telescope schemes offer little advantage over this
+Telescope algorithm.
+The simple lottery, however, offers the smallest communication complexity mu
+and fastest aggregation time.
+It can likely be the top choice for applications where certificate size is not
+important.
+We should decide if we want to provide both unweighted and weighted versions, or
+simply the weighted versions.
+At least for the centralized setting, unweighted versions can be concretely more
+efficient due to no lottery and less complexity.
+For the decentralized setting, an unweighted version could be more efficient as
+well due to not needing to sample the binomial distribution.
+
+What hash function to use for ALBA?
+Ideally, the user should be able to provide his own hash function.
```

Comment:
How would this done practically? If the library is to be embedded in other languages (eg. Haskell), this might be quite difficult

Reply:
Perhaps we could bind to functions with a set directly?

Reply:
Hard to say without knowledge of Haskell and interoperability. One option, though not perfect, is to create a new library in Rust that instantiates all types with concrete types and use that in Haskell. Resolving for now.